python/lsst/ts/ledprojector/ledprojector_csc.py

```diff
@@ -0,0 +1,336 @@
+        self.config = config
+
+#from our perspective, there can only be the 'led model' as one object, since the only failure point is the labjack
+#the labjack then connects, effectively one way, to the LEDs, so from this objects perspective we only have on, off, and no failure states
+#the failure state lies in the labjack not connecting/responding
+    async def handle_summary_state(self):
+        if self.disabled_or_enabled:
+            if not self.led_connected:
+                try:
+                    await self.connect_led()
+                except Exception as e:
+                    return await self.fault(
+                        code=ErrorCode.led_ERROR,
+                        report=f"Could not connect to the led: {e!r}",
+                    )
+            self.should_be_connected = True
+        elif self.summary_state == salobj.State.FAULT:
+            if self.led_model and self.led_model.led_was_on:
+                # Turn off the led, if connected
+                if self.led_connected:
+                    self.log.warning(
+                        "Going to fault while connected to the led controller; "
+                        "trying to turn the led off immediately."
+                    )
+                    try:
+                        await self.led_model.turn_led_off(
+                            force=True, wait=False, reason="CSC is going to FAULT state"
+                        )
+                    except Exception as e:
+                        self.log.error(
+                            f"Failed to turn led off; please turn it off manually: {e!r}"
+                        )
+                else:
+                    self.log.warning(
+                        "Going to fault and not connect to the led controller; "
+                        "please turn the led off manually."
+                    )
+        else:
+            self.should_be_connected = False
+            if self.led_connected and self.led_model.led_was_on:
+                try:
+                    await self.led_model.turn_led_off(
+                        force=True,
+                        wait=False,
+                        reason=f"CSC is going to state {self.summary_state!r}",
+                    )
+                except Exception as e:
+                    self.log.warning(
+                        f"Going to state {self.summary_state!r} but failed to turn off led: {e!r}; "
+                        "please turn it off manually."
+                    )
+            await self.disconnect_led()
+
+
+    async def connect_led(self):
+        """Connect to the led controller LabJack and get status.
+
+        Raises
+        ------
+        asyncio.TimeoutError
+            If it takes longer than self.config.led.connect_timeout
+        """
+        await self.disconnect_led()
+        if self.led_model is None:
+            self.led_model = LEDModel(
+                config=self.config.led,
+                csc=self,
```

Comment:
Please, don't pass the CSC to the LEDModel class.

Reply:
So this depends on whether Joseph will actually set_write events or telemetry in the model or not. I know Russell prefered that the CSC would be passed through to the controller if the controller was going to set a property that would be published via SAL anyway and then set_write the event/telemetry directly thereby reducing duplication. You can look at the GIS or CBP as an example of that.